File: src/_cpp/TowerDefenseBase/support/AutoPlayer.cpp
#include "AutoPlayer.h"
#include "MapLayer.h"
#include "GameGS.h"
#include "ScoreCounter.h"
#include "tower.h"
#include "UserData.h"
NS_CC_BEGIN


#define SCHEDULE(func, interval) Director::getInstance()->getScheduler()->scheduleSelector( schedule_selector( AutoPlayer::func ), this, interval, false )
#define UNSCHEDULE() Director::getInstance()->getScheduler()->unscheduleAllForTarget( this )
MapLayer* getMap()
{
	auto scene = Director::getInstance()->getRunningScene();
	auto node = scene->getChildByName( "maplayer" );
	return dynamic_cast<MapLayer*>(node);
}

GameGS* getGame()
{
	auto scene = Director::getInstance()->getRunningScene();
	auto node = scene->getChildByName( "gamelayer" );
	return dynamic_cast<GameGS*>(node);
}

AutoPlayer*inst( nullptr );

AutoPlayer::AutoPlayer()
: _mode(GameMode::normal)
{
	inst = this;
}

AutoPlayer::~AutoPlayer( )
{
	inst = nullptr;
}

AutoPlayer* AutoPlayer::getInstance()
{
	return inst;
}

bool AutoPlayer::init( bool runCurrentLevel, bool onlyOnceLevel, float rate, bool withoutDefeat )
{
	time_t t;
	time( &t );
	srand( t );

	_runCurrentLevel = runCurrentLevel;
	_onlyOnceLevel = onlyOnceLevel;
	_withoutDefeat = withoutDefeat;

	_currentLevel = runCurrentLevel ? getMap()->_selectedLevelIndex : 0;
	if( _currentLevel == -1 )
		_currentLevel = 0;

	Director::getInstance()->setTimeRate( rate );

	SCHEDULE( state_selectLevel, 1 );
	
	return true;
}

void AutoPlayer::state_selectLevel( float dt )
{
	if( getMap() )
	{
		_currentTowerForBuild = "";
		ScoreCounter::shared().setMoney( kScoreFuel, 100, false );
		getMap()->runLevel( _currentLevel, _mode );
		UNSCHEDULE();
		SCHEDULE( state_waitLoading, 1 );
	}
}

void AutoPlayer::state_waitLoading( float dt )
{
	if( getGame() )
	{
		UNSCHEDULE();
		SCHEDULE( state_play, 1.5f );
	}
}

void AutoPlayer::state_play( float dt )
{
	if( !GameGS::getInstance() )
	{
		UNSCHEDULE();
		release();
		return;
	}
	if( GameGS::getInstance()->getGameBoard().checkGameFinished() )
	{
		UNSCHEDULE();
		SCHEDULE( state_victory, 1 );
		return;
	}

	if( GameGS::getInstance()->getGameBoard().isGameStarted() == false )
	{
		int score = ScoreCounter::shared().getMoney( kScoreLevel );
		ScoreCounter::shared().addMoney( kScoreLevel, score / 2, false );

		if( _withoutDefeat )
			ScoreCounter::shared().setMoney( kScoreHealth, 99999, false );
	}

	auto icon = getNodeByPath<WaveIcon>( GameGS::getInstance(), "interface/waveicon" );
	bool click = GameGS::getInstance()->getGameBoard().isGameStarted() == false;
	if( icon )
	{
		click = click || icon->_elapsed > icon->_duration / 2.5;
	}
	if( click && icon )
	{
		icon->on_click( nullptr );
	}

	auto places = GameGS::getInstance()->getTowerPlaces();
	int scores = ScoreCounter::shared().getMoney( kScoreLevel );
	if( places.empty() == false )
	{
		if( _currentTowerForBuild.empty() )
		{
			while( true )
			{
				std::list<std::string> towers;
				mlTowersInfo::shared().fetch( towers );
				int index = rand() % towers.size();
				while( index-- )
					towers.pop_front();
				_currentTowerForBuild = towers.front();
				if( _previuosTowerForBuild != _currentTowerForBuild && _previuosTowerForBuild2 != _currentTowerForBuild )
					break;
			}
		}

		int cost = mlTowersInfo::shared().getCost( _currentTowerForBuild, 1 );
		if( cost <= scores )
		{
			TowerPlace::Pointer place;
			int min( 0 );
			float min_d( 9999999 );
			auto decorations = GameGS::getInstance()->getDecorations( "base_point" );
			if( decorations.empty() == false )
			{
				auto decor = decorations[rand() % decorations.size()];
				int index( 0 );
				for( auto place : places )
				{
					float d = place->getPosition().getDistance( decor->getPosition() );
					if( d < min_d )
					{
						min_d = d;
						min = index;
					}
					++index;
				}
				place = places[min];
			}
			else
			{
				int index = rand() % places.size();
				place = places[index];
			}


			if( place )
			{
				GameGS::getInstance()->setSelectedTowerPlaces( place );
				_towers.push_back( GameGS::getInstance()->getGameBoard().createTower( _currentTowerForBuild ) );
				_previuosTowerForBuild2 = _previuosTowerForBuild;
				_previuosTowerForBuild = _currentTowerForBuild;
				_currentTowerForBuild.clear();
			}
		}
	}
	else if( _towers.empty() == false )
	{
		auto towerForUpgrade = _towers[rand() % _towers.size()];

		int level = towerForUpgrade->getLevel();
		int cost = mlTowersInfo::shared().getCost( towerForUpgrade->getName(), level + 1 );
		if( cost <= scores )
		{
			auto tower = GameGS::getInstance()->getGameBoard().upgradeTower( towerForUpgrade );
			if( tower )
			{
				_towerUpgraded.push_back( tower );
				auto iter = std::find( _towers.begin(), _towers.end(), towerForUpgrade );
				if( iter != _towers.end() )
					_towers.erase( iter );
			}
		}

		if( _towers.empty() )
		{
			_towers = _towerUpgraded;
			_towerUpgraded.clear();
		}
	}
}

void AutoPlayer::state_victory( float dt )
{
	if( ScoreCounter::shared().getMoney( kScoreHealth ) > 0 )
		++_currentLevel;
	
	_currentTowerForBuild.clear();
	_previuosTowerForBuild.clear();
	_previuosTowerForBuild2.clear();
	_towers.clear();
	_towerUpgraded.clear();

	Director::getInstance()->popScene();
	UNSCHEDULE();
	SCHEDULE( state_waitMap, 1 );
}

void AutoPlayer::state_waitMap( float dt )
{
	if( getMap() )
	{
		UNSCHEDULE();
		SCHEDULE( state_loop, 1 );
	}
}

void AutoPlayer::state_loop( float dt )
{
	UNSCHEDULE();
	if( !_onlyOnceLevel )
	{
		SCHEDULE( state_selectLevel, 1 );
	}
	else
	{
		release();
	}
}

#undef SCHEDULE
#undef UNSCHEDULE

NS_CC_END

File: src/_eu/support/AutoPlayer.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.AutoPlayer = cc.Class.extend(
{
    /** @type {int} */ _currentLevel : null,
    /** @type {string} */ _currentTowerForBuild : null,
    /** @type {string} */ _previuosTowerForBuild : null,
    /** @type {string} */ _previuosTowerForBuild2 : null,
    /** @type {Array<EU.Unit>} */ _towers : null,
    /** @type {Array<EU.Unit>} */ _towerUpgraded : null,

    /** @type {bool} */ _runCurrentLevel : null,
    /** @type {bool} */ _onlyOnceLevel : null,
    /** @type {bool} */ _withoutDefeat : null,
    /** @type {EU.GameMode} */ _mode : null,

    setGameMode: function( /**EU.GameMode*/ mode ) { this._mode = mode; },

    SCHEDULE: function(func, interval) {
        "use strict";
        cc.director.getScheduler().schedule( func , this, interval, false )
    },

    UNSCHEDULE: function(){
        "use strict";
        cc.director.getScheduler().unscheduleAllForTarget( this );
    },

    getMap: function()
    {
        var scene = cc.director.getRunningScene();
        var node = scene.getChildByName( "maplayer" );
        return node && node.__MapLayer ? node : null;
    },

    getGame: function()
    {
        var scene = cc.director.getRunningScene();
        var node = scene.getChildByName( "gamelayer" );
        return node.__GameGS ? node : null;
    },

    onExit: function()
    {
        this.sharedDirector = null;
    },

    ctor: function(runCurrentLevel, onlyOnceLevel, rate, withoutDefeat )
    {
        this._mode = EU.GameMode.normal;
        this._towers = [];
        this._towerUpgraded = [];

        //var t = Date.now();
        //srand( t );

        this._runCurrentLevel = runCurrentLevel;
        this._onlyOnceLevel = onlyOnceLevel;
        this._withoutDefeat = withoutDefeat;

        this._currentLevel = runCurrentLevel ? this.getMap().selectedLevelIndex : 0;
        if( this._currentLevel == -1 )
            this._currentLevel = 0;

        //TODO: check if setTimeScale can replace setTimeRate
        cc.director.getScheduler().setTimeScale( rate );

        this.SCHEDULE( this.state_selectLevel, 1 );

        return true;
    },

    state_selectLevel: function( /**@type {var} */ dt )
    {
        if( this.getMap() )
        {
            this._currentTowerForBuild = "";
            EU.ScoreCounter.setMoney( EU.kScoreFuel, 100, false );
            this.getMap().runLevel( this._currentLevel, this._mode );
            this.UNSCHEDULE();
            this.SCHEDULE( this.state_waitLoading, 1 );
        }
    },

    state_waitLoading: function( /**@type {var} */ dt )
    {
        if( this.getGame() )
        {
            this.UNSCHEDULE();
            this.SCHEDULE( this.state_play, 1.5 );
        }
    },

    state_play: function( /**@type {var} */ dt )
    {
        if( !EU.GameGSInstance )
        {
            this.UNSCHEDULE();
            release();
            return;
        }
        if( EU.GameGSInstance.getGameBoard().checkGameFinished() )
        {
            this.UNSCHEDULE();
            this.SCHEDULE( this.state_victory, 1 );
            return;
        }

        if( EU.GameGSInstance.getGameBoard().isGameStarted() == false )
        {
           var score = EU.ScoreCounter.getMoney( EU.kScoreLevel );
            EU.ScoreCounter.addMoney( EU.kScoreLevel, score / 2, false );

            if( this._withoutDefeat )
                EU.ScoreCounter.setMoney( EU.kScoreHealth, 99999, false );
        }

        var icon = EU.Common.getNodeByPath( EU.GameGSInstance, "interface/waveicon" );
        var click = EU.GameGSInstance.getGameBoard().isGameStarted() == false;
        if( icon )
        {
            click = click || icon._elapsed > icon._duration / 2.5;
        }
        if( click && icon )
        {
            icon.on_click( null );
        }

        var places = EU.GameGSInstance.getTowerPlaces();
        var scores = EU.ScoreCounter.getMoney( EU.kScoreLevel );
        if( places.length > 0 )
        {
            if( EU.xmlLoader.stringIsEmpty(this._currentTowerForBuild) )
            {
                while( true )
                {
                    var towers = [];
                    EU.mlTowersInfo.fetch( towers );
                    var index = cc.rand() % towers.length;
                    while( index-- )
                        towers.shift();
                    this._currentTowerForBuild = towers[0];
                    if( this._previuosTowerForBuild != this._currentTowerForBuild && this._previuosTowerForBuild2 != this._currentTowerForBuild )
                        break;
                }
            }

            var cost = EU.mlTowersInfo.getCost( this._currentTowerForBuild, 1 );
            if( cost <= scores )
            {
               /**EU.TowerPlace*/
               var place;
               var min = 0 ;
                var min_d = 9999999 ;
                var decorations = EU.GameGSInstance.getDecorations( "base_point" );
                if( decorations.length > 0 )
                {
                    var decor = decorations[cc.rand() % decorations.length];
                    var index = 0 ;
                    for (var i = 0; i < places.length; i++) {
                        var place = places[i];
                        var d = cc.pDistance(place.getPosition(), decor.getPosition() );
                        if( d < min_d )
                        {
                            min_d = d;
                            min = index;
                        }
                        ++index;
                    }
                    place = places[min];
                }
                else
                {
                   var index = cc.rand() % places.length;
                    place = places[index];
                }


                if( place )
                {
                    EU.GameGSInstance.setSelectedTowerPlaces( place );
                    this._towers.push( EU.GameGSInstance.getGameBoard().createTower( this._currentTowerForBuild ) );
                    this._previuosTowerForBuild2 = this._previuosTowerForBuild;
                    this._previuosTowerForBuild = this._currentTowerForBuild;
                    this._currentTowerForBuild = "";
                }
            }
        }
        else if( this._towers.length > 0 )
        {
            var towerForUpgrade = this._towers[cc.rand() % this._towers.length];

           var level = towerForUpgrade._level;
           var cost = EU.mlTowersInfo.getCost( towerForUpgrade.getName(), level + 1 );
            if( cost <= scores )
            {
                var tower = EU.GameGSInstance.getGameBoard().upgradeTower( towerForUpgrade );
                if( tower )
                {
                    this._towerUpgraded.push( tower );
                    var index = this._towers.indexOf(towerForUpgrade );
                    if( index >= 0)
                        this._towers.splice( index , 1 );
                }
            }

            if( this._towers.length == 0 )
            {
                this._towers = this._towerUpgraded;
                this._towerUpgraded = [];
            }
        }
    },

    state_victory: function( /**@type {var} */ dt )
    {
        if( EU.ScoreCounter.getMoney( EU.kScoreHealth ) > 0 )
            ++this._currentLevel;

        this._currentTowerForBuild = "";
        this._previuosTowerForBuild = "";
        this._previuosTowerForBuild2 = "";
        this._towers = [];
        this._towerUpgraded = [];

        cc.director.popScene();
        this.UNSCHEDULE();
        this.SCHEDULE( this.state_waitMap, 1 );
    },

    state_waitMap: function( /**@type {var} */ dt )
    {
        if( this.getMap() )
        {
            this.UNSCHEDULE();
            this.SCHEDULE( this.state_loop, 1 );
        }
    },

    state_loop: function( /**@type {var} */ dt )
    {
        this.UNSCHEDULE();
        if( !this._onlyOnceLevel )
        {
            this.SCHEDULE( this.state_selectLevel, 1 );
        }
        else
        {
            release();
        }
    }

});


EU.AutoPlayer.sharedInstance = null;
EU.AutoPlayer.firstUseInstance = true;

EU.AutoPlayer._getInstance = function () {
    if (EU.AutoPlayer.firstUseInstance) {
        EU.AutoPlayer.firstUseInstance = false;
        EU.AutoPlayer.sharedInstance = new EU.AutoPlayer(null, null, 1, false);
    }
    return EU.AutoPlayer.sharedInstance;
};
//EU.autoPlayer = EU.AutoPlayer._getInstance();